File: 24.swap-nodes-in-pairs.java
/*
 * [24] Swap Nodes in Pairs
 *
 * https://leetcode.com/problems/swap-nodes-in-pairs/description/
 *
 * algorithms
 * Medium (38.97%)
 * Total Accepted:    202.1K
 * Total Submissions: 518.5K
 * Testcase Example:  '[]'
 *
 * 
 * Given a linked list, swap every two adjacent nodes and return its head.
 * 
 * 
 * 
 * For example,
 * Given 1->2->3->4, you should return the list as 2->1->4->3.
 * 
 * 
 * 
 * Your algorithm should use only constant space. You may not modify the values
 * in the list, only nodes itself can be changed.
 * 
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
		ListNode dummyHead = new ListNode(0);
		ListNode prev = dummyHead;
		dummyHead.next = head;
		while (head != null && head.next != null) {
			ListNode next = head.next;
			ListNode nextHead = head.next.next;
			prev.next = next;
			next.next = head;
			head.next = nextHead;
			prev = head;
			head = nextHead;
		}
		return dummyHead.next;		
    }
}

File: go/24.swap-nodes-in-pairs.go
/*
 * [24] Swap Nodes in Pairs
 *
 * https://leetcode.com/problems/swap-nodes-in-pairs/description/
 *
 * algorithms
 * Medium (39.17%)
 * Total Accepted:    212.7K
 * Total Submissions: 542.5K
 * Testcase Example:  '[1,2,3,4]'
 *
 * Given a linked list, swap every two adjacent nodes and return its head.
 * 
 * Example:
 * 
 * 
 * Given 1->2->3->4, you should return the list as 2->1->4->3.
 * 
 * Note:
 * 
 * 
 * Your algorithm should use only constant extra space.
 * You may not modify the values in the list's nodes, only nodes itself may be
 * changed.
 * 
 * 
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	p := dummy
	for p.Next != nil && p.Next.Next != nil {
		first := p.Next
		second := first.Next
		next := second.Next
		second.Next = first
		first.Next = next
		p.Next = second
		p = first
	}
	return dummy.Next
}
